### PR TITLE
add img matrix infra x kubectl

### LINKED ISSUE #37: Provide Trimmed-Down Images per Infrastructure and Kubernetes Version
**What would you like to be added**:
Today the images:
- are overly large (2x or 4x as large as the original first images; e.g. 200MB vs. 100MB for the plain toolbelt and 436MB vs. 116MB for the toolbelt with the `aws` CLI image)
- contain all infrastructures CLIs even though an operator needs usually only one (for a particular cluster)
- contains a 1.17 `kubectl`, which can be only used for clusters within the [skew policy](https://kubernetes.io/docs/setup/release/version-skew-policy/#kubectl), so is mostly the wrong one for the currently supported Kubernetes versions in Gardener

**Why is this needed**:
The current images are too large, not specific enough (which adds to the first problem and results in longer startup times and 2-4x higher network usage and costs), and features an outdated `kubectl` that may fail/will be useless for clusters of a particular version.

**What do you suggest**:
Let's:
1. First understand why the images became so large and what we can do to cut them down and get them smaller again (fewer components, cleanup, different base OS if it's not too disruptive)
2. Build a matrix of images (infrastructure x `kubectl` version), so that a user and the dashboard can pick the right one (gardener/dashboard#842) and of course also the big one (for local usage)

### PR DESCRIPTION
**Pull request by jfortin-sap**
**What this PR does / why we need it**:
Add img matrix infra x kubectl to speed up ops toolbelt load times
**Which issue(s) this PR fixes**:
Fixes #37 

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Add img matrix infra x kubectl to speed up ops toolbelt load times
```

### CODE REVIEW

File: .ci/build [hunk 1]
```diff
@@ -12,8 +12,17 @@ if [[ -z "${GENERATED_DOCKERFILES_PATH}" ]]; then
     mkdir -p "${GENERATED_DOCKERFILES_PATH}"
 fi
 
+cat > "${SOURCE_PATH}"/dockerfile-configs/kubectl-components.yaml << EOF
+    - curl:
+      - name: kubectl
+        version: v1.20.5
```
Comment (petersutter) [inline .ci/build:18]: take the first line of `k8s_versions ` which should be the latest version

File: .ci/build [hunk 2]
```diff
@@ -22,46 +31,63 @@ fi
...
+        --dockerfile "${GENERATED_DOCKERFILES_PATH}"/ops-toolbelt-openstack-k$i.dockerfile
+
+    "${SOURCE_PATH}"/generator/generate-dockerfile.py \
+        --dockerfile-configs "${SOURCE_PATH}"/dockerfile-configs/common-components.yaml \
+        --dockerfile-configs "${SOURCE_PATH}"/dockerfile-configs/aliyun-components.yaml \
+        --dockerfile-configs "${SOURCE_PATH}"/dockerfile-configs/gardenctl-components.yaml \
+        --dockerfile-configs "${SOURCE_PATH}"/dockerfile-configs/kubectl-components.yaml \
+        --from-image ubuntu:18.04 \
+        --title "gardener shell" \
+        --dockerfile "${GENERATED_DOCKERFILES_PATH}"/ops-toolbelt-aliyun-k$i.dockerfile
```
Comment (petersutter) [inline .ci/build:92]: you could also have a loop for the iaas components

File: .ci/build_pipeline_definitions [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            inputs:
+              steps:
+                build: ~
+          ops-toolbelt-aliyun-k$i:
+            registry: 'gcr-readwrite'
+            image: 'eu.gcr.io/gardener-project/gardener/ops-toolbelt-aliyun-k$i'
+            dockerfile: ops-toolbelt-aliyun-k$i.dockerfile
+            inputs:
+              steps:
+                build: ~
```
Comment (petersutter) [inline .ci/build_pipeline_definitions:74]: same here, loop over iaas names

File: .ci/build_pipeline_definitions [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+EOF
+
+for i in `cat "${SOURCE_PATH}"/.ci/k8s_versions`
+do
+    cat >> "${SOURCE_PATH}"/.ci/pipeline_definitions << EOF
+          ops-toolbelt-aws-k$i:
+            registry: 'gcr-readwrite'
+            image: 'eu.gcr.io/gardener-project/gardener/ops-toolbelt-aws-k$i'
```
Comment (petersutter) [inline .ci/build_pipeline_definitions:42]: I don't think we should have the patch version in the image name. Major and minor version should be sufficient

File: .ci/build [hunk 5]
```diff
@@ -22,46 +31,63 @@ fi
...
+done
```
Comment (petersutter) [inline .ci/build:93]: I assume the `kubectl-components.yaml` file is only used temporarily to generate the dockerfile, right? Shouldn't this file be cleaned up afterwards? It anyhow only contains the contents of the last run of the loop

**Review by petersutter — CHANGES_REQUESTED** [review]
please also document what needs to be done in order to support new kubectl versions (or drop deprecated versions), which scripts to run and to commit..

File: .ci/build_pipeline_definitions [hunk 6]
```diff
@@ -34,45 +34,21 @@ ops-toolbelt:
...
+    do
+
+      cat >> "${SOURCE_PATH}"/.ci/pipeline_definitions << EOF
+            ops-toolbelt-$iaas_provider-k$kv_short:
+              registry: 'gcr-readwrite'
+              image: 'eu.gcr.io/gardener-project/gardener/ops-toolbelt-$iaas_provider-k$kv_short'
+              dockerfile: ops-toolbelt-$iaas_provider-k$kv_short.dockerfile
+              inputs:
+                steps:
+                  build: ~
```
Comment (petersutter) [inline .ci/build_pipeline_definitions:49]: the indenting is not correct. After the fix, please re-run the script to fix `.ci/pipeline_definitions`
```suggestion
          ops-toolbelt-$iaas_provider-k$kv_short:
            registry: 'gcr-readwrite'
            image: 'eu.gcr.io/gardener-project/gardener/ops-toolbelt-$iaas_provider-k$kv_short'
            dockerfile: ops-toolbelt-$iaas_provider-k$kv_short.dockerfile
            inputs:
              steps:
                build: ~
```

**Review by petersutter — CHANGES_REQUESTED** [review]

**Comment by petersutter** [comment]
@jfortin-sap this point is still open https://github.com/gardener/ops-toolbelt/pull/50#pullrequestreview-655018628

**Comment by jfortin-sap** [comment]
@petersutter PR updated

File: README.md [hunk 7]
```diff
@@ -73,6 +73,19 @@ To build all pre-configured images run:
 $ .ci/build
 ```
 
+To update pre-configured images supported k8s versions run then commit:
+```bash
+$ gardenctl target garden live
+$ gardenctl kubectl get cloudprofile -o json | jq -r '.items[].spec.kubernetes.versions[] | select(.classification=="supported") | .version' |sort -r |uniq > .ci/k8s_versions
+$ .ci/build_pipeline_definitions
```
Comment (petersutter) [inline README.md:80]:
```suggestion
gardenctl target garden <landscape>
gardenctl kubectl get cloudprofile -o json | jq -r '.items[].spec.kubernetes.versions[] | select(.classification=="supported") | .version' |sort -r |uniq > .ci/k8s_versions
.ci/build_pipeline_definitions
```

File: README.md [hunk 8]
```diff
@@ -73,6 +73,19 @@ To build all pre-configured images run:
...
+```
+
+To update pre-configured images supported iaas providers run then commit:
+```bash
+$ <editor> .ci/iaas-providers
+$ .ci/build_pipeline_definitions
```
Comment (petersutter) [inline README.md:86]:
```suggestion
<editor> .ci/iaas-providers # to update the list
.ci/build_pipeline_definitions
```

File: README.md (hunk 7 above)
Comment (petersutter) [reply README.md:80]: The command to update the `k8s_versions` file should make sure that only the latest patch version of a minor version is added.
e.g. the versions are
```
1.20.5
1.19.9
1.19.8
```

only 
```
1.20.5
1.19.9
```
should be added

File: README.md [hunk 9]
```diff
@@ -73,6 +73,19 @@ To build all pre-configured images run:
 $ .ci/build
 ```
 
+To update pre-configured images supported k8s versions run then commit:
```
Comment (petersutter) [inline README.md:76]:
```suggestion
To update pre-configured images supported k8s versions run the following commands and then commit the resulting changes:
```

File: README.md [hunk 10]
```diff
@@ -73,6 +73,19 @@ To build all pre-configured images run:
...
+```bash
+$ gardenctl target garden live
+$ gardenctl kubectl get cloudprofile -o json | jq -r '.items[].spec.kubernetes.versions[] | select(.classification=="supported") | .version' |sort -r |uniq > .ci/k8s_versions
+$ .ci/build_pipeline_definitions
+```
+
+To update pre-configured images supported iaas providers run then commit:
```
Comment (petersutter) [inline README.md:83]:
```suggestion
To update pre-configured images supported iaas providers run the following commands and then commit the resulting changes:
```

File: README.md (hunk 7 above)
Comment (petersutter) [reply README.md:80]: What about the deprecated minor versions? You can still create clusters for them

**Review by petersutter — CHANGES_REQUESTED** [review]

**Comment by jfortin-sap** [comment]
@petersutter Updated, ready to merge?

**Comment by jfortin-sap** [comment]
@petersutter ready to merge?

**Review by petersutter — APPROVED** [review]
/lgtm

**Review by petersutter — APPROVED** [review]
/lgtm

**Pull request merged**